Fix visible state of buttons in the Relation Editor Widget

signedav opened the pull request:
`mZoomToFeatureButton` visibility is set on several places and it used to override each other. Same for `mSaveEditsButton` and `mAddFeatureGeometryButton`.

To keep it stable and not being dependent on the order of the setting, it store the configured visibility of buttons in `mButtonsVisibility` and calls `upadteButtons` from `setVisibleButtons`, `setRelation`, `initDualView` and `setEditorContext`.

For `mZoomToFeatureButton` it checks if the flag is set, the editorContexts mapCanvas exists and if the referenced layer of `relation` / the referencing layer of the `nmRelation` is spacial
For `mAddFeatureGeometryButton` it checks the same, but additionally for the existence of `mEditorContext.cadDockWidget()`.

For `mSaveEditsButton` and `mToggleEditingButton` it checks has the member `mLayerInSameTransactionGroup`. This is additionally set correctly according if the parent AND the child are in this group (or with many-to-many relations additionally the "other" parent layer as well).

fixes #37718

signedav committed:
mZoomToFeatureButton visibility is set on several places. To keep it stable and not beeing dependant on the order or the setting it stores mChildIsSpacial and mButtons (the configuration).

On setVisibleButtons it does only set mZoomToFeatureButton visible, if the referenced layer of relation / the referencing layer of the nmRelation is spacial and the flag is set.
On setRelation it does only set mZoomToFeatureButton visible, if the referenced layer of relation / the referencing layer of the nmRelation is spacial and the flag of the config is set.
On updateButtons it does only set mZoomToFeatureButton visible, if the referenced layer of relation / the referencing layer of the nmRelation is spacial and the flag is set and the mapCanvas of the context is existing.

fixes #37718

signedav committed:
On one to many relation it should check the referencing layer for being spatial instead of the referenced layer

signedav committed:
use the spatial state ofthe child instead of checking the geometry on every buttonUpdate

signedav committed:
remove the visibility set of mZoomToFeatureButton on button update since the config or the spatial state do not change here

3nids commented on src/gui/qgsrelationeditorwidget.cpp:
```diff
@@ -959,14 +945,15 @@ bool QgsRelationEditorWidget::showSaveChildEditsButton() const
+  mAddFeatureGeometryButton->setVisible( mButtons.testFlag( QgsAttributeEditorRelation::Button::AddChildFeature ) );
+  mDuplicateFeatureButton->setVisible( mButtons.testFlag( QgsAttributeEditorRelation::Button::DuplicateChildFeature ) );
+  mDeleteFeatureButton->setVisible( mButtons.testFlag( QgsAttributeEditorRelation::Button::DeleteChildFeature ) );
+  mZoomToFeatureButton->setVisible( mButtons.testFlag( QgsAttributeEditorRelation::Button::ZoomToChildFeature ) && mChildIsSpatial );
```
```suggestion
  mZoomToFeatureButton->setVisible( mButtons.testFlag( QgsAttributeEditorRelation::Button::ZoomToChildFeature ) && mChildIsSpatial && mEditorContext.mapCanvas() );
```

3nids reviewed (approved):
Looks much cleaner!
What about the check for the map canvas, can it be introduced as proposed?

signedav committed:
handle all button visibility in updateButtons()

signedav committed:
set mSaveEditButtons according to the visibility of mToggleEditingButton depending on the transactionGroups AND the buttons visibility setting

signedav committed:
better name for buttons visibility setting stored in the widget

3nids commented on src/gui/qgsrelationeditorwidget.cpp:
```diff
@@ -427,27 +420,18 @@ void QgsRelationEditorWidget::updateButtons()
+
+  mLinkFeatureButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::Link ) );
+  mUnlinkFeatureButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::Unlink ) );
+  mSaveEditsButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::SaveChildEdits ) && mToggleEditingButton->isVisible() );
```
```suggestion
  mSaveEditsButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::SaveChildEdits ) && editable );
```

sounds better no?

3nids commented on src/gui/qgsrelationeditorwidget.cpp:
```diff
@@ -427,27 +420,18 @@ void QgsRelationEditorWidget::updateButtons()
+
+  mLinkFeatureButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::Link ) );
+  mUnlinkFeatureButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::Unlink ) );
+  mSaveEditsButton->setVisible( mButtonsVisibility.testFlag( QgsAttributeEditorRelation::Button::SaveChildEdits ) && mToggleEditingButton->isVisible() );
```
can't we use another condition that the button visibility?

signedav replied:
One possibility would be to check if the referencing layer is in the `transactionGroup` but I don't like that it iterates here since it's not really something that changed.

Another possibility would be to set the status in something like `mLayerInTransactionGroup = true` or `mLayerEditable = false` here https://github.com/qgis/QGIS/pull/38678/files#diff-52237b5949ffab9a9049c00ed3a713e1L332 and then check the member variable in the `setVisible`.

What do you think?

3nids replied:
yes, let's go for the member, it will make the logic more explicit.

3nids reviewed (approved):
nice!

signedav committed:
use member to keep state if related layer is in transaction group to set toggleEditing and SaveEdits button visibility

check the "other parent" in transaction group depending on if it's a many-to-many cardinality or one-to-many

fix that it sets toggle-edit and save button to invisible if child layer is in transaction group (even if it's not in the same transaction group like the parent)

Means:
on one-to-many, we check if parent and child are in the same transaction group:
parent-layer (relation->referencedLayer)
child-layer (relation->referencingLayer)

on many-to-many, we check if parent and child and the other parent are in the same transaction group:
parent-layer (relation->referencedLayer)
child-layer / relation-layer (relation->referencingLayer)
other parent layer (nmrelation->referencedLayer)

signedav committed:
default value for mButtonsVisibility on initialization since it's used when creating a widget with no settings (because of an autogenerated form)

signedav commented:
Thanks @3nids
